runtime/platform/platform.h

```diff
@@ -126,12 +142,134 @@ void et_pal_emit_log_message(
  *     et_pal_free().
  */
 void* et_pal_allocate(size_t size) ET_INTERNAL_PLATFORM_WEAKNESS;
+typedef void* (*et_pal_allocate_t)(size_t size);
 
 /**
  * Frees memory allocated by et_pal_allocate().
  *
  * @param[in] ptr Pointer to memory to free. May be nullptr.
  */
 void et_pal_free(void* ptr) ET_INTERNAL_PLATFORM_WEAKNESS;
+typedef void (*et_pal_free_t)(void* ptr);
 
 } // extern "C"
+
+namespace executorch {
+namespace runtime {
+
+/**
+ * Table of pointers to platform abstraction layer functions.
+ */
+struct pal_table {
+  pal_table() = default;
+
+  pal_table(
+      et_pal_emit_log_message_t emit_log_message,
+      const char* source_filename);
+
+  pal_table(
+      et_pal_init_t init,
+      et_pal_abort_t abort,
+      et_pal_current_ticks_t current_ticks,
+      et_pal_ticks_to_ns_multiplier_t ticks_to_ns_multiplier,
+      et_pal_emit_log_message_t emit_log_message,
+      et_pal_allocate_t allocate,
+      et_pal_free_t free,
+      const char* source_filename);
+
+  et_pal_init_t init = nullptr;
+  et_pal_abort_t abort = nullptr;
+  et_pal_current_ticks_t current_ticks = nullptr;
+  et_pal_ticks_to_ns_multiplier_t ticks_to_ns_multiplier = nullptr;
+  et_pal_emit_log_message_t emit_log_message = nullptr;
+  et_pal_allocate_t allocate = nullptr;
+  et_pal_free_t free = nullptr;
+
+  // An optional metadata field, indicating the name of the source
+  // file that registered the PAL implementation.
+  const char* source_filename;
+};
+
+/**
+ * Override the PAL functions with user implementations. Any null entries in the
+ * table are unchanged and will keep the default implementation.
+ *
+ * Returns true if the registration was successful, false otherwise.
+ */
+bool register_pal(pal_table impl);
```

Comment:
This API is returning bool primarily because using Error would introduce a circular header dependency and there's no current way for this call to fail. I'm open to switching to error, but would require a little bit of a header refactor.